[B5] Web Apps

orangejenny opened the pull request:
## Product Description
Turn on bootstrap 5 for web apps. Side by side comparisons are [here](https://docs.google.com/document/d/1aw_F8A5u73CXaqNT0NEq0p3idh9PoVmqdErKRa8mN1g/edit), but the high level summary is: not much changed. Colors are a little brighter, borders are a little lighter, buttons are a little bigger.

## Technical Summary
Review by commit. Most commits are small, except the cleanup commits, which are mechanical. Most of this work has been merged already, what remains in this PR is three major changes:

* Updating the python views to use the B5 decorators and templates
* Javascript changes that either depend on B5 libraries (bootstrap js or tempus dominus) or were annoying to wrap behind a `window.USE_BOOTSTRAP5` check.
* Deleting the B3 templates and diff files.

## Safety Assurance

### Safety story
High risk, as this is a high visibility area, there are a lot of changes (including changes made in previous PRs that will go live when this is merged), and our test coverage at the UI level is limited. We'll be going through the same QA/safety approach as for the web apps requirejs migration.

### Automated test coverage

Not a whole lot.

### QA Plan

https://dimagi.atlassian.net/browse/QA-6550

### Rollback instructions

This can be rolled back if needed. Due to its size, it may get difficult to automatically roll back relatively soon, although it looks like there hasn't been that much activity in the web apps front end in the past few weeks.

Sentry can be monitored for javascript errors using [this filter](https://dimagi.sentry.io/issues/?environment=production&project=136860&query=is%3Aunresolved+staticfiles&referrer=issue-list&statsPeriod=7d).

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

orangejenny committed:
Updated javascript dependencies and added TODOs

This is basically the result of running the bootstrap migration tool.

orangejenny committed:
Switched views to B5

orangejenny committed:
Updated modals

orangejenny committed:
Updated spinners for clickable icons

orangejenny committed:
Fixed grid layouts not displaying icons inline

This finishes up https://github.com/dimagi/commcare-hq/pull/34647/commits/a43a5276b2b6f36d21694a7d9e2628165f221946

orangejenny committed:
Migrated to tempus dominus

- dateTimePickerTooltips are already handled by by defaultTranslations in the HQ tempus dominus module
- Several callbacks/handlers were no longer needed
- It's annoying that showing the clear button is handled through initial page data, but updating that is out of scope

orangejenny committed:
Updated CSS classes in javascript

The migration tool doesn't check js files for CSS class names,
and some of these aren't greppable anyway.

orangejenny committed:
Finished updating map in case list logic

This completes changes that were started in 845d9ffc8b3fda932c7057b0bf11a87ce3361eaf

orangejenny committed:
Removed B3 files

orangejenny committed:
Updated diff config

The LESS stays in here because it's still used by integration/base_cloudcare_template.html

orangejenny committed:
Moved B5 files back into default positions

orangejenny committed:
Bootstrap 5 Migration - Rebuilt diffs

orangejenny committed:
Marked cloudcare migration as complete

orangejenny committed:
Removed USE_BOOSTRAP5 in js

orangejenny committed:
Removed bootstrap switch detritus

orangejenny committed:
Updated settings OQPS event

Followup for https://github.com/dimagi/commcare-hq/commit/ee098afdbee6a00a05b0ca42e6b58686a80081c2

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/utils.js:
```diff
@@ -433,7 +437,9 @@ hqDefine('cloudcare/js/utils', [
         let date = moment(selectedTime, timeFormat);
-        $el.datetimepicker({
+        // todo B5: move to tempus dominus
+        //hqTempusDominus.createTimePicker($el);
```
why is tempusDominus commented out too? are we doing neither right now?

MartinRiese replied:
I see you did that in a later commit

MartinRiese commented on corehq/apps/hqwebapp/tests/data/bootstrap5_diffs/stylesheets/formplayer-common_grid.formplayer-common_grid.style.diff.txt:
```diff
@@ -0,0 +1,7 @@
+--- 
++++ 
```
why are these not unsplit?

orangejenny committed:
Bootstrap 5 Migration - Rebuilt diffs

orangejenny committed:
Updated test_update_gruntfile

orangejenny committed:
Fixed lint errors in Gruntfile.js

orangejenny committed:
Updated MochaView to know which apps have been migrated to B5

The alternative is to recreate the bootstrap5 directories for cloudcare
mocha templates, which is differently ugly. This will need to be updated
as apps are migrated to B5, and then removed when the entire platform
is migrated.

orangejenny committed:
Fixed search icon alignment in app preview

Oddly, this is an app preview issue, not a small screen size issue.

orangejenny committed:
Reduced size of confirmation modal buttons

orangejenny committed:
Updated pagination to allow for wrapping in app preview

orangejenny committed:
Moved styles to center breadcrumb separators to also apply to app preview

orangejenny committed:
"Bootstrap 5 Migration - Rebuilt diffs"

orangejenny committed:
Removed focusout handlers, set up time questions to initialize to current time

orangejenny committed:
Increased important of select2 clear icon position

orangejenny committed:
Fixed checkbox styling for combined checkboxes (QA-6621)

orangejenny committed:
Restored background color for menu badges (QA-6622)

orangejenny committed:
Fixed datepicker event (QA-6604)

orangejenny committed:
Added tempus dominus compatibility in a few stray places

orangejenny committed:
Updated case search errors to display property

My least favorite thing about this migration is that .invalid-feedback items
only display if they have a sibling with the .is-invalid class. This is a change from
B3, where error messages were styled based on being in a container that used .has-error.

orangejenny committed:
Restored asterisk for require case search fields (QA-6600)

orangejenny committed:
Updated a few styling rules dependent on classes that are no longer used

orangejenny committed:
Fixed error styling for select questions (QA-6627)

This is a bug in https://github.com/dimagi/commcare-hq/pull/34631/
It's causing choices to display error styling if there's an erroranywhere in the form.

orangejenny committed:
Removed contradictory repeat group styles

orangejenny committed:
Updated more styles relating to B3 classes no longer being used

orangejenny commented:
Force pushed to address problem with merging in a conflict resolution branch that had been based on an earlier version of this branch. Did not change any individual commits, just removed the duplicate commits added by that merge.

orangejenny committed:
Fixed incorrect check for is-invalid in select choices

This should match the other change in afb5663c3ab601e803337ae64cd1f2944253ba3b

orangejenny committed:
Merge branch 'master' into jls/b5-web-apps

orangejenny committed:
Imported print styles from B3

orangejenny committed:
Updated print styles to not limit content width

orangejenny committed:
Updated col-* rules in print-general.scss

orangejenny committed:
Updated col-* rules in form.scss

orangejenny committed:
"Bootstrap 5 Migration - Rebuilt diffs"

orangejenny committed:
Added (back) datepicker.scss

This largely restores B3's preview_app/datepicker.less, just updating the selector's
class name and adding an override for transform.

orangejenny committed:
"Bootstrap 5 Migration - Rebuilt diffs"

orangejenny committed:
Updated stray references to .panel

orangejenny committed:
Removed add_grouop.html

I missed this while rebasing on master.

orangejenny committed:
Fixed rendering and saving of OQPS setting

orangejenny committed:
Fixed padding in form error UI

This .col-md-12 isn't necessary, as alerts already take up all available width,
and it's messing with the content's padding.

orangejenny committed:
Removed unnecessary wrapper containers in debugger's Evaluate XPath area

orangejenny committed:
Removed spacing wrappers around new repeat button

These don't do anything.

orangejenny committed:
Cleaned up file upload input

The col.*classes add extra right-left padding that misaligns the input with the label.

orangejenny committed:
Added gap to question tiles

This addresses the spacing issue for repeat groups in question tiles that 4c0c688 took care of in B3.

orangejenny committed:
Simplified question.html

This make for better consistency when trying to align questions, labels, and buttons that use question tiles.

orangejenny committed:
Fixed lint errors

orangejenny committed:
Fixed OQPS setting so it can be turned off, not just on

orangejenny committed:
Merge branch 'master' into jls/b5-web-apps

orangejenny committed:
Fixed display of multiselect checkboxes in case tiles

orangejenny committed:
Merge branch 'master' into jls/b5-web-apps

orangejenny committed:
Updated app preview to show required asterisks

orangejenny committed:
"Bootstrap 5 Migration - Rebuilt diffs"

orangejenny committed:
Merge branch 'master' into jls/b5-web-apps

orangejenny committed:
Removed bottom border from multiselect checkbox cells, to match other case list cells

orangejenny committed:
Reduced padding for Login As banner on app preview

orangejenny committed:
Updated dropdowns to use .form-select instead of .form-control

orangejenny committed:
"Bootstrap 5 Migration - Rebuilt diffs"

orangejenny committed:
Fixed refine search button missing at certain screen widths

orangejenny committed:
Fixed early wrapping of text in collapsible group headers (SAAS-15619)

MartinRiese committed:
Merge branch 'master' into jls/b5-web-apps

orangejenny committed:
Revert "Adjusted responsive design for header on case search screen" (QA-6725)

This reverts commit f946ca3fc67d0980e19fae33b2d8ab8d75bc7c8b.

orangejenny committed:
Merge branch 'jls/b5-web-apps' of https://github.com/dimagi/commcare-hq into jls/b5-web-apps

orangejenny committed:
Updated date and time widgets to display default values

orangejenny committed:
Updated date picker to close when date is selected

This behavior is not added to the time picker or the date range picker since
neither of those are single-click interactions.

orangejenny committed:
Added error handling for unrecognizable date/time answers

orangejenny committed:
Added date formats

This makes tempus dominus more forgiving about users entering single-digit
months and years, which would previously be rejected as invalid input.

MartinRiese commented on corehq/apps/hqwebapp/static/hqwebapp/js/tempus_dominus.js:
```diff
@@ -20,7 +20,7 @@ hqDefine("hqwebapp/js/tempus_dominus", [
 
     let createDatePicker = function (el, options) {
-        return new tempusDominus.TempusDominus(el, _addDefaultOptions(options, {
+        let picker = new tempusDominus.TempusDominus(el, _addDefaultOptions(options, {
```
Is the return not needed anymore? I would have expected a `return picker` below.

MartinRiese replied:
never mind. I see it now

MartinRiese reviewed (commented):
started backwards because I am not sure how far I got before. Stopped at 15e030b3105e3bc178185a185f4124ec6ce3ca88

MartinRiese commented on corehq/apps/hqwebapp/static/cloudcare/scss/formplayer-common/form.scss:
```diff
@@ -34,10 +34,3 @@ legend {
     padding-top: 20px;
 }
-
-.question-tile-row {
```
👍

MartinRiese reviewed (commented):
got to 5858f7c6511f8f01a95477bcb579d3b0b247e8f0

orangejenny committed:
Fixed offset class (QA-6738)

orangejenny committed:
Fixed alignment of checkbox questions with lengthy labels (QA-6739)

orangejenny committed:
Updated setting of time value (QA-6735)

This doesn't need to be formatted.

MartinRiese reviewed (commented):
new changes look good

MartinRiese committed:
Merge branch 'master' into jls/b5-web-apps

MartinRiese committed:
SAAS-15620: Add margin to h1-h6

* Bootstrap 5 removes the top margin of h1-h6. This affects the
  collapsible group header layout
* Add a class to add an equal margin for h1-h6 top and bottom
* Use it for group headers

MartinRiese committed:
"Bootstrap 5 Migration - Rebuilt diffs"

nospame commented on corehq/apps/hqwebapp/static/cloudcare/scss/formplayer-webapp/form.scss:
```diff
@@ -263,3 +263,10 @@ $group-indent: 15px;
   bottom: 35px;
 }
+
+.mtb-05em-for-hs {
```
B5 convention would have this prefixed `.my-` where `y` means top and bottom.
https://github.com/twbs/bootstrap/blob/v5.0.2/site/content/docs/5.0/utilities/spacing.md?plain=1#L33
The `05em-for-hs` itself feels verbose but it's descriptive; I don't think it _needs_ to change and I don't have any immediate suggestions.

MartinRiese committed:
Update left over reference from $model to $loginModal

MartinRiese committed:
Remove module-column-name class

In the context of the login as user table the
module-column-name class only set the position to relative.
Since neither the element itself nor it children set a
position offset the only effect it had was resetting the z-index
which interfered with the focus highlight

MartinRiese committed:
Fix linting

MartinRiese committed:
"Bootstrap 5 Migration - Rebuilt diffs"

MartinRiese committed:
Use y instead of tb for top and bottom

Just following bootstrap convention here

MartinRiese committed:
Replace order classes with insert at right spot

In Bootstrap 3 we were able to have and intermediate tag as parent
of the apps. This is not allowed anymore in Bootstrap 5. The order
classes previously used only change the visual order, not the tab
order. Adding a hidden reference div and inserting the apps before
it results in the right order that also keeps the tab order in place.

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/apps/views.js:
```diff
@@ -112,6 +112,13 @@ hqDefine("cloudcare/js/formplayer/apps/views", [
 
+        // Cannot append at the end of the parent because there are
+        // special grid items.
+        attachHtml: function attachHtml(els, $container) {
+            let childElement = $container.find("#put-apps-here");
+            $container[0].insertBefore(els, childElement[0]);
+        },
```
I don't love this solution but I looked at it for a while and couldn't come up with anything better.

MartinRiese replied:
same here

MartinRiese committed:
Move relative position to direct parent

The relative position is needed to place the hover circle
around the delete/audio icon correctly. But it interfers
with the focus outline of the row.

* Remove it from the row related class
* Add bootstrap class to add the property to the direct
  parent of the icon
* Add separate column for icons to improve the code structure.

MartinRiese committed:
"Bootstrap 5 Migration - Rebuilt diffs"

MartinRiese committed:
Fix css for sscs

* Button bar should be there independent of screen size
* break point for SSCS is lg not md

MartinRiese committed:
"Bootstrap 5 Migration - Rebuilt diffs"

MartinRiese committed:
Hide case-list-menu-header in preview

minhaminha committed:
Merge branch 'master' into jls/b5-web-apps

MartinRiese committed:
Commenting out invalid validation regex

MartinRiese committed:
Merge branch 'jls/b5-web-apps' of github.com:dimagi/commcare-hq into jls/b5-web-apps

nospame commented on corehq/apps/cloudcare/templates/cloudcare/partials/users/user_row.html:
```diff
@@ -4,7 +4,7 @@
     </div>
   </td>
-  <td class="module-column">
+  <td class="module-column module-column-name">
```
The `module-column-name` class gets added here but it looks like the css class was removed, is that right? Or does it serve some other purpose as well?

MartinRiese replied:
I had removed it to get rid of the relative positioning that caused issues. With this commit I am adding it back in because it missing caused some other issue in app preview. Instead I removed the property from the css class and added it to the direct parent of the icon tags that needed the relative position.

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/utils.js:
```diff
@@ -380,7 +380,7 @@ hqDefine('cloudcare/js/utils', [
 
         $el.attr("placeholder", dateFormat);
-        $el.attr("pattern", "[0-9\-/]+");   // eslint-disable-line no-useless-escape
+        // $el.attr("pattern", "[0-9\-/]+");   // eslint-disable-line no-useless-escape
```
What was this used for, and why comment out rather than delete?

MartinRiese replied:
I think it is used for validation but it caused and error with each validation because it is malformed regex. I was not sure what the correct regex should be. So I commented it out for now.

nospame replied:
@esoergel do you have any context on what this is meant to be/do? I know you did work with the datepicker a while back and it looks like one of your commits put this pattern here.

MartinRiese replied:
@nospame changed it to `$el.attr("pattern", "\\d{1,2}/\\d{1,2}/\\d{4}");` since the format is hard coded to 'M/D/YYYY'.

esoergel replied:
I think this is meant to be "integers, dashes, or slashes" - IIRC, users were sometimes typing out "July 1st" or whatever, and this would mark that as an error.  The existing datepicker actually accepts string input in a variety of formats, defined here.  I believe it starts from left to right and uses the first format that matches:
https://github.com/dimagi/commcare-hq/blob/1a70089d74f4c7c367799ef83e7eddcd2739340a/corehq/apps/cloudcare/static/cloudcare/js/utils.js#L376-L376

esoergel replied:
Some more context and discussion on that here: https://github.com/dimagi/commcare-hq/pull/32738

MartinRiese replied:
that makes more sense. In the HTML it should end up looking like this `pattern="[0-9\-\/]+"` because we the dash and forward slash are inside the [] we need to escape them. But since we are in JS land we need two \ each to get there. so `$el.attr("pattern", "[0-9\\-\\/]+");`

nospame reviewed (commented):
Just reviewed the changes since last review of "new changes look good". Couple clarifying questions, looks good on the whole.

nospame reviewed (approved):
Think it'll be good to find an answer to https://github.com/dimagi/commcare-hq/pull/34682#discussion_r1672605060 but not considering it blocking. Approving the changes made here by @MartinRiese in the time @orangejenny has been offline.

MartinRiese committed:
Fix regex for date input pattern

MartinRiese committed:
Fix page header alignment

MartinRiese committed:
No hover effect on nested query fields

Was not a thing with B3 either

MartinRiese committed:
"Bootstrap 5 Migration - Rebuilt diffs"

nospame commented on corehq/apps/hqwebapp/static/cloudcare/scss/formplayer-webapp/content.scss:
```diff
@@ -18,12 +18,19 @@ body {
     text-transform: uppercase;
     color: $cc-neutral-mid;
-    padding-left: 1.5rem;
+    padding-left: 12px;
```
Is there a relative padding that could be used here instead (e.g. `1rem`)?

esoergel replied:
There's also `$spacer`, though I'm not certain when you'd use that over `rem` (I believe `$spacer: 2rem`)

MartinRiese replied:
I used px because because `.module-column-icon` does too and I did not want to mix two units.

esoergel replied:
I don't think this needs to block, but another option would be to use a common reference variable, if the two items need to line up.  Can you explain the connection between `.page-header > h1` and `.module-column-icon`?  I see one is set to 12px and the other to 20px, is there an 8px element in the mix somewhere too?

MartinRiese committed:
Merge branch 'master' into jls/b5-web-apps

esoergel reviewed (approved):
Approving commits since Evan's last approval
